mslice/widgets/ipythonconsole/ipython_widget.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import (absolute_import, division,
+                        print_function)
+
+import inspect
+import threading
+import types
+import warnings
+
+from PyQt4 import QtGui
+
+# IPython monkey patches the  pygments.lexer.RegexLexer.get_tokens_unprocessed method
+# and breaks Sphinx when running within MantidPlot.
+# We store the original method definition here on the pygments module before importing IPython
+from pygments.lexer import RegexLexer
+# Monkeypatch!
+RegexLexer.get_tokens_unprocessed_unpatched = RegexLexer.get_tokens_unprocessed
+
+# Ignore Jupyter/IPython deprecation warnings that we can't do anything about
+warnings.filterwarnings('ignore', category=DeprecationWarning, module='IPython.*')
+warnings.filterwarnings('ignore', category=DeprecationWarning, module='ipykernel.*')
+warnings.filterwarnings('ignore', category=DeprecationWarning, module='jupyter_client.*')
+warnings.filterwarnings('ignore', category=DeprecationWarning, module='qtconsole.*')
+del warnings
+
+try:
+    # Later versions of Qtconsole are part of Jupyter
+    from qtconsole.rich_jupyter_widget import RichJupyterWidget as RichIPythonWidget
+    from qtconsole.inprocess import QtInProcessKernelManager
+except ImportError:
+    # fix for bug with old ipython versions
+    import mslice.external.ipython as _ # noqa
+
+    from IPython.qt.console.rich_ipython_widget import RichIPythonWidget
+    from IPython.qt.inprocess import QtInProcessKernelManager
+
+def our_run_code(self, code_obj, result=None):
+    """ Method with which we replace the run_code method of IPython's InteractiveShell class.
+        It calls the original method (renamed to ipython_run_code) on a separate thread
+        so that we can avoid locking up the whole of MantidPlot while a command runs.
+
+        Parameters
+        ----------
+        code_obj : code object
+          A compiled code object, to be executed
+        result : ExecutionResult, optional
+          An object to store exceptions that occur during execution.
+        Returns
+        -------
+        False : Always, as it doesn't seem to matter.
+    """
+
+    t = threading.Thread()
+    #ipython 3.0 introduces a third argument named result
+    nargs = len(inspect.getargspec(self.ipython_run_code).args)
+    if (nargs == 3):
+        t = threading.Thread(target=self.ipython_run_code, args=[code_obj,result])
+    else:
+        t = threading.Thread(target=self.ipython_run_code, args=[code_obj])
+    t.start()
+    while t.is_alive():
+        QtGui.QApplication.processEvents()
+    # We don't capture the return value of the ipython_run_code method but as far as I can tell
+    #   it doesn't make any difference what's returned
+    return 0
+
+
+class IPythonWidget(RichIPythonWidget):
+    """ Extends IPython's qt widget to include setting up and in-process kernel as well as the
+        Mantid environment, plus our trick to avoid blocking the event loop while processing commands.
```

Comment:
Please remove the part about the Mantid environment as that's not applicable here.